### ISSUE TITLE
[Bug]: DeepSeek-V4 W8A8 Concurrency Failure: Output Corruption and Malformed begin of sentence with MTP Enabled

### ISSUE THREAD

**Issue by w1027-1** [vllm-project/vllm-ascend#9170]
### Your current environment

vLLM Version: 0.18.0+empty (Editable install)

Model: DeepSeek-V4 (DSV4)

Quantization: W8A8

Attention Backend: FlashAttention (Ascend)

Speculative Config: {"num_speculative_tokens": 1, "method": "deepseek_mtp"}

Hardware: Ascend 910B (or equivalent)

### 🐛 Describe the bug

A critical execution error has been observed when running concurrent requests on DeepSeek-V4 (W8A8 quantized) with DeepSeek MTP and FlashAttention on Ascend NPUs.

The Phenomenon:
During concurrent load, the model's output suddenly injects a <｜begin of sentence｜> (or malformed begin?of?sentence) string, and the response terminates immediately after this injection. This prevents the model from completing its response and results in truncated, nonsensical outputs.

eg:
{"id":"chatcmpl-279128cb9fcfda5c6d7ba6699f7c605b","object":"chat.completion.chunk","created":1778151413,"model":"deepseek_v4","choices":[{"index":0,"delta":{"role":"assistant","content":""},"logprobs":null,"finish_reason":null}],"prompt_token_ids":null}

data: {"id":"chatcmpl-279128cb9fcfda5c6d7ba6699f7c605b","object":"chat.completion.chunk","created":1778151413,"model":"deepseek_v4","choices":[{"index":0,"delta":{"content":"<｜begin▁of▁sentence｜>"},"logprobs":null,"finish_reason":null,"token_ids":null}]}

data: {"id":"chatcmpl-279128cb9fcfda5c6d7ba6699f7c605b","object":"chat.completion.chunk","created":1778151413,"model":"deepseek_v4","choices":[{"index":0,"delta":{},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null}]}

data: [DONE]

Single Request: Works normally without any unexpected tokens.

Concurrent Requests (MTP Enabled): Frequent early termination preceded by BOS token injection.

Concurrent Requests (MTP Disabled): 100% stable; no early termination or BOS injection observed.

**Comment by weijinqian0** [comment]
@nomewang @GDzhu01 Plz see this issue.

**Comment by onkevin** [comment]
我遇到了相同的BOS问题，硬件是910C，部署按照这个方案A3单机部署，https://docs.vllm.ai/projects/ascend/en/v0.18.0/tutorials/models/DeepSeek-V4-Flash.html
关闭MTP时就没有问题了。

**Comment by nicholas9698** [comment]
@onkevin 我也遇到了相同的BOS问题（910C），您模型用的是Eco-Tech/DeepSeek-V4-Flash-w8a8-mtp吗？我测试了关闭MTP还是存在问题。

**Comment by willrone** [comment]
我也遇到了一样的问题，关闭MTP的话会无限输出<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>直到上下文撑爆

**Comment by kennylbj** [comment]
同样问题 atlas 800 a3, 开了mtp，begin▁of▁sentence，关了 mtp ，模型回答出现乱码

**Comment by Mitchellax** [comment]
800I A2，MTP1 / 关闭MTP，流式输出 / 非流式输出。均出现了<｜begin▁of▁sentence｜>泄漏。部分情况下会在输出 `<｜begin▁of▁sentence｜>cherrypick` 后直接截断，部分情况会在开始时输出 `<｜begin▁of▁sentence｜>` 后继续正常输出。

**Comment by LL1024LL** [comment]
同样的问题，a3单机，开了mtp 1

**Comment by GDzhu01** [comment]
Please send us the request to reproduce the problem and the minimal service scenario.

**Comment by kennylbj** [comment]
Any progress on this thread?

**Comment by ejunered** [comment]
tp>=4时出现这个情况。 tp>=4,把VLLM_ASCEND_ENABLE_FUSED_MC2关掉也没问题。

**Comment by ppppangu** [comment]
Hit this exact symptom on **DeepSeek-V4-Flash W8A8 + MTP** (`num_speculative_tokens=1`, `deepseek_mtp`), 910C A3, TP4×DP4 + expert-parallel (16 ranks), vLLM 0.18 + vllm-ascend: intermittent **empty / malformed `<｜begin▁of▁sentence｜>` first response**. **MTP-off is 100% reliable; MTP-on intermittently corrupts the first decode step**, and `VLLM_ASCEND_ENABLE_FUSED_MC2=0` eliminates it.

Traced (source-level) to the fused `torch.ops._C_ascend.dispatch_ffn_combine` op being fed the **tiny/ragged spec-decode token shape**. The rejection sampler is clean (it always emits the target-argmax token); the corruption is in the fused MoE op's output, upstream of sampling:

1. **`unpermute/moe_token_unpermute_tiling.h:30-34`** — sole writer of the per-token output (called on-device from `dispatch_ffn_combine_kernel.hpp:926`). Per-core token count via integer division, no min-1 guard: when a rank's local `outTokens < coreNum` (AIV cores), `tokens_core_length = outTokens/coreNum = 0` → `tokens_splited_length = min(0,600) = 0` → `tokens_splited_num = 0/0`, `tokens_splited_remain = 0%0` (integer divide-by-zero). The undefined values bound the output-emitting loop (`moe_token_unpermute.h:181-187`), so a garbage/zero number of output rows is written.
2. The op **never zero-initializes** `params.ptrOutput` / combine staging (everything sized by the fixed `max_output_size=65536`), so under-filled rows on the tiny batch are read from stale HBM → NaN/Inf/degenerate logits → empty / BOS-only first response.
3. Corroborating, same tiny-shape regime: W8A8 dynamic-quant scale `= max(|row|)/127` then `Div(x, scale)` with no epsilon (`moe_init_routing_quant_v2/moe_v2_fullload_dynamic_quant.h:244-256` + identical clones) → Inf/NaN on a degenerate routed row; and the int8 `expandedX` matmul-input buffer is never memset with its dirty-data guard compiled out under `#ifdef __CCE_KT_TEST__` (`moe_v2_src_to_dst_with_capacity.h:134-139`).

Matmul/swiglu correctly skip empty groups (`currentM>0` guards), so the corruption is specifically the output-assembly + uninitialized-staging path on the out-of-distribution tiny shape; normal large/dense decode never hits the 0/0 (tokens ≫ cores).

A proper kernel fix would add a min-1 guard in the unpermute tiling and zero-init the output staging for under-filled batches — but that touches the hot non-spec MoE path and needs on-NPU validation. As a safe, side-effect-free control-plane mitigation, PR #9983 disables fused MC2 under speculative decoding (the existing `not is_draft_model` guard misses the **target** verify forward, which runs with `is_draft_model=False`).

**Comment by ppppangu** [comment]
**Follow-up: this is reachable without speculative decoding — #9983 mitigates the MTP trigger but does not close #9170.**

Reading the `dispatch_ffn_combine` AscendC source (CANN Open Software), the corruption is driven by per-rank token count and expert-routing distribution — neither is a function of `speculative_config`. Plain high-concurrency decode/prefill reaches the same defects.

**Dominant non-spec trigger (source-provable).** Under concurrent batching, expert imbalance can drive a local expert's token count to 0 → `currentM == 0` → the per-expert GMM is skipped (`dispatch_ffn_combine_kernel.hpp`, `if (currentM > 0)`) → those rows of the **never-zeroed** output (`out = torch.empty_like(...)` in `FusedMC2CommImpl.fused_experts`, `ops/fused_moe/moe_comm_method.py`) are read back as stale HBM → NaN/garbage first-step logits → empty / malformed-BOS. Because the device allocator sometimes hands back already-zeroed pages, this is intermittent — which matches the "concurrency-dependent" framing of this issue.

**Secondary trigger (source-provable).** `MoeTokenUnpermuteTiling` computes `tokens_core_length = outTokens / (coreNum/2)` then `tokens_splited_num = tokens_core_length / tokens_splited_length` (`unpermute/moe_token_unpermute_tiling.h`). When per-rank `outTokens < coreNum/2` (the AIC core count, fixed per SoC), `tokens_core_length == 0` → integer `0/0`. Again a function of token count, not MTP.

**Candidate fixes.**
- *Cheap, broad, Python:* `torch.empty_like` → `torch.zeros_like` for the combine output (`moe_comm_method.py`) so skipped/under-filled rows become a deterministic zero (also the correct combine identity) instead of stale HBM. Helps both the spec and non-spec paths. Posted as a small PR.
- *Kernel (full root-fix):* memset the combine staging + output up to the actual routed row count before the per-expert GMM/combine; clamp the unpermute tiling divisor (`max(1, …)`); add an epsilon to the W8A8 dynamic-quant scale (`moe_init_routing_quant_v2/moe_v2_fullload_dynamic_quant.h`, the `max(|row|)/127`); and add a per-call epoch to the `tokenPerExpert` cross-rank sentinel (set then `gm_signal_wait_until_ne(...)`, reset only at the end) so concurrent batches on one EP rank cannot observe a stale sentinel.

**Confidence.** The skipped-GMM/stale-output and the `0/0`-tiling conditions, and their independence from `speculative_config`, are source-provable. Needs on-NPU confirmation: whether a given workload actually drives a rank's M below the AIC core count / a local expert to 0 tokens; whether concurrent `dispatch_ffn_combine` launches truly overlap on one stream (the `tokenPerExpert` sentinel has no epoch, so it assumes single-in-flight); and whether the `empty_like` rows are observably NaN vs benign-zero on the live allocator.

PR #9983 remains a safe control-plane mitigation for the MTP/speculative trigger (non-spec bit-identical); the kernel defects above still need a fix for the general concurrency case.

**Comment by WillWillWong** [comment]
At present, the problem still remains as to how to deal with it.

**Comment by wangzhishenghw** [comment]
**Automated Response**

This deepseek-v4 BOS token leakage / output corruption issue has been actively discussed in comments. @ppppangu provided detailed kernel-level analysis of the fused MC2 op path and proposed a fix direction. @GDzhu01 please review the analysis and coordinate the fix. Model series confirmed: **DSv4**.
